File: go/goals_test.go
package kairos

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestGoalsList(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)

		response := map[string]interface{}{
			"goals": []map[string]interface{}{
				{
					"id":          "goal-1",
					"team_id":     "team-1",
					"title":       "Q1 Objectives",
					"description": "Goals for Q1",
					"status":      "active",
					"created_by":  "user-1",
					"created_at":  time.Now().UTC().Format(time.RFC3339),
					"updated_at":  time.Now().UTC().Format(time.RFC3339),
				},
			},
			"total":     1,
			"hasMore":   false,
			"limit":     10,
			"count":     1,
			"offset":    0,
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	client, _ := New("test-key", WithBaseURL(server.URL))
	goals, pagination, err := client.Goals.List(context.Background(), nil)

	if err != nil {
		t.Fatalf("List failed: %v", err)
	}

	if len(goals) != 1 {
		t.Errorf("Expected 1 goal, got %d", len(goals))
	}

	if goals[0].Title != "Q1 Objectives" {
		t.Errorf("Expected title 'Q1 Objectives', got '%s'", goals[0].Title)
	}

	if pagination.Total != 1 {
		t.Errorf("Expected total 1, got %d", pagination.Total)
	}
}

func TestGoalsGet(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)

		response := map[string]interface{}{
			"goal": map[string]interface{}{
				"id":          "goal-1",
				"team_id":     "team-1",
				"title":       "Q1 Objectives",
				"description": "Goals for Q1",
				"status":      "active",
				"created_by":  "user-1",
				"created_at":  time.Now().UTC().Format(time.RFC3339),
				"updated_at":  time.Now().UTC().Format(time.RFC3339),
			},
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	client, _ := New("test-key", WithBaseURL(server.URL))
	goal, err := client.Goals.Get(context.Background(), "goal-1")

	if err != nil {
		t.Fatalf("Get failed: %v", err)
	}

	if goal.ID != "goal-1" {
		t.Errorf("Expected ID 'goal-1', got '%s'", goal.ID)
	}

	if goal.Title != "Q1 Objectives" {
		t.Errorf("Expected title 'Q1 Objectives', got '%s'", goal.Title)
	}
}

func TestGoalsCreate(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			w.WriteHeader(http.StatusMethodNotAllowed)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)

		response := map[string]interface{}{
			"goal": map[string]interface{}{
				"id":         "goal-2",
				"team_id":    "team-1",
				"title":      "New Goal",
				"status":     "active",
				"created_by": "user-1",
				"created_at": time.Now().UTC().Format(time.RFC3339),
				"updated_at": time.Now().UTC().Format(time.RFC3339),
			},
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	client, _ := New("test-key", WithBaseURL(server.URL))
	input := CreateGoalInput{Title: "New Goal"}
	goal, err := client.Goals.Create(context.Background(), input)

	if err != nil {
		t.Fatalf("Create failed: %v", err)
	}

	if goal.Title != "New Goal" {
		t.Errorf("Expected title 'New Goal', got '%s'", goal.Title)
	}
}

func TestGoalsUpdate(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPatch {
			w.WriteHeader(http.StatusMethodNotAllowed)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)

		response := map[string]interface{}{
			"goal": map[string]interface{}{
				"id":     "goal-1",
				"title":  "Updated Goal",
				"status": "completed",
			},
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	client, _ := New("test-key", WithBaseURL(server.URL))
	status := "completed"
	input := UpdateGoalInput{Status: &status}
	goal, err := client.Goals.Update(context.Background(), "goal-1", input)

	if err != nil {
		t.Fatalf("Update failed: %v", err)
	}

	if goal.Status != "completed" {
		t.Errorf("Expected status 'completed', got '%s'", goal.Status)
	}
}

func TestGoalsListTasks(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)

		response := map[string]interface{}{
			"tasks": []map[string]interface{}{
				{
					"id":         "task-1",
					"team_id":    "team-1",
					"goal_id":    "goal-1",
					"title":      "Task 1",
					"status":     "open",
					"created_by": "user-1",
					"created_at": time.Now().UTC().Format(time.RFC3339),
					"updated_at": time.Now().UTC().Format(time.RFC3339),
				},
			},
			"total":     1,
			"hasMore":   false,
			"limit":     10,
			"count":     1,
			"offset":    0,
		}
		json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	client, _ := New("test-key", WithBaseURL(server.URL))
	tasks, pagination, err := client.Goals.ListTasks(context.Background(), "goal-1", nil)

	if err != nil {
		t.Fatalf("ListTasks failed: %v", err)
	}

	if len(tasks) != 1 {
		t.Errorf("Expected 1 task, got %d", len(tasks))
	}

	if pagination.Total != 1 {
		t.Errorf("Expected total 1, got %d", pagination.Total)
	}
}

File: typescript/tests/goals.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { Kairos } from '../src/client';
import type { Goal, CreateGoalInput } from '../src/types';

describe('GoalsResource', () => {
  let client: Kairos;

  beforeEach(() => {
    process.env.KAIROS_API_KEY = 'kairos_sk_test_key_12345';
  });

  describe('list()', () => {
    it('should fetch goals with pagination', async () => {
      const mockGoals: Goal[] = [
        {
          id: 'goal_1',
          team_id: 'team_1',
          title: 'Q2 Objectives',
          description: 'Quarterly objectives for Q2',
          status: 'active',
          progress: 45,
          due_date: '2026-06-30',
          created_by: 'user_1',
          created_at: '2026-04-05T00:00:00Z',
          updated_at: '2026-04-05T00:00:00Z',
        },
      ];

      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({
              goals: mockGoals,
              count: 1,
              total: 1,
              hasMore: false,
              limit: 10,
              offset: 0,
            }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      const result = await client.goals.list();

      expect(result.data).toEqual(mockGoals);
      expect(result.pagination.total).toBe(1);
    });

    it('should filter goals by status', async () => {
      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({
              goals: [],
              count: 0,
              total: 0,
              hasMore: false,
              limit: 10,
              offset: 0,
            }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      await client.goals.list({ status: 'active' });

      const callUrl = (global.fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('status=active');
    });
  });

  describe('get()', () => {
    it('should fetch a single goal by ID', async () => {
      const mockGoal: Goal = {
        id: 'goal_1',
        team_id: 'team_1',
        title: 'Q2 Objectives',
        description: 'Quarterly objectives',
        status: 'active',
        progress: 45,
        due_date: '2026-06-30',
        created_by: 'user_1',
        created_at: '2026-04-05T00:00:00Z',
        updated_at: '2026-04-05T00:00:00Z',
      };

      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({ goal: mockGoal }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      const result = await client.goals.get('goal_1');

      expect(result).toEqual(mockGoal);
    });
  });

  describe('create()', () => {
    it('should create a new goal', async () => {
      const input: CreateGoalInput = {
        title: 'New Goal',
        description: 'Goal description',
        due_date: '2026-06-30',
      };

      const createdGoal: Goal = {
        id: 'goal_new',
        team_id: 'team_1',
        ...input,
        status: 'active',
        progress: 0,
        created_by: 'user_1',
        created_at: '2026-04-05T00:00:00Z',
        updated_at: '2026-04-05T00:00:00Z',
      };

      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({ goal: createdGoal }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      const result = await client.goals.create(input);

      expect(result).toEqual(createdGoal);
      expect(global.fetch).toHaveBeenCalledWith(
        expect.stringContaining('/goals'),
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify(input),
        }),
      );
    });
  });

  describe('update()', () => {
    it('should update an existing goal', async () => {
      const updateInput = {
        status: 'completed' as const,
      };

      const updatedGoal: Goal = {
        id: 'goal_1',
        team_id: 'team_1',
        title: 'Q2 Objectives',
        description: 'Quarterly objectives',
        status: 'completed',
        progress: 100,
        due_date: '2026-06-30',
        created_by: 'user_1',
        created_at: '2026-04-05T00:00:00Z',
        updated_at: '2026-04-05T01:00:00Z',
      };

      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({ goal: updatedGoal }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      const result = await client.goals.update('goal_1', updateInput);

      expect(result.status).toBe('completed');
      expect(result.progress).toBe(100);
    });
  });

  describe('listTasks()', () => {
    it('should fetch tasks under a goal', async () => {
      const mockTasks = [
        {
          id: 'task_1',
          team_id: 'team_1',
          goal_id: 'goal_1',
          parent_task_id: null,
          title: 'Task 1',
          description: null,
          type: 'task' as const,
          status: 'to_do' as const,
          priority: 'high' as const,
          assigned_to: null,
          estimated_hours: null,
          actual_hours: null,
          due_date: null,
          completed_at: null,
          created_by: 'user_1',
          created_at: '2026-04-05T00:00:00Z',
          updated_at: '2026-04-05T00:00:00Z',
        },
      ];

      global.fetch = vi.fn(() =>
        Promise.resolve(
          new Response(
            JSON.stringify({
              tasks: mockTasks,
              count: 1,
              total: 1,
              hasMore: false,
              limit: 10,
              offset: 0,
            }),
            {
              status: 200,
              headers: {
                'content-type': 'application/json',
                'X-RateLimit-Limit-Minute': '100',
                'X-RateLimit-Remaining-Minute': '99',
                'X-RateLimit-Limit-Hour': '1000',
                'X-RateLimit-Remaining-Hour': '999',
                'X-RateLimit-Reset': '1712304000',
              },
            },
          ),
        ),
      );

      client = new Kairos();
      const result = await client.goals.listTasks('goal_1');

      expect(result.data).toEqual(mockTasks);
    });
  });
});
